xen/arch/arm/vscmi.c

```diff
@@ -0,0 +1,356 @@
+            break;
+        }
+
+        if ( level < 0 || level >= PERF_OPP_COUNT) {
+            gprintk(XENLOG_WARNING, "vscmi: requested opp is out of bounds: %d\n", level);
+            write_32(data->msg_payload, SCMI_ERR_PARAMS);
+            data->length = sizeof(uint32_t) * 2;
+            break;
+        }
+
+        current->domain->vcpu[domain]->arch.opp = level;
+        /* TODO: Check the return value */
+        notifier_call_chain(&vscmi_chain, 0, current->domain->vcpu[domain], NULL);
+
+        write_32(data->msg_payload, SCMI_SUCCESS);
+        data->length = sizeof(uint32_t) * 2;
+
+        break;
+    }
+    case PERF_LEVEL_GET:
+    {
+        uint32_t domain = le32_to_cpu(*(__le32*)data->msg_payload);
+
+        if ( domain >= current->domain->max_vcpus ) {
+            write_32(data->msg_payload, SCMI_ERR_RANGE);
+            data->length = sizeof(uint32_t) * 2;
+            break;
+        }
+
+        write_32(data->msg_payload, SCMI_SUCCESS);
+        write_32(data->msg_payload + 4, current->domain->vcpu[domain]->arch.opp + 1);
+        data->length = sizeof(uint32_t) * 3;
+
+        break;
+    }
+    default:
+        write_32(data->msg_payload, SCMI_ERR_SUPPORT);
+        data->length = sizeof(uint32_t) * 2;
+        break;
+    }
+}
+
+bool vscmi_handle_call(struct cpu_user_regs *regs)
+{
+    struct scmi_shared_mem *data;
+    uint32_t hdr;
+    int res;
+
+    if ( !current->domain->arch.scmi_base_pg )
+    {
+        printk(XENLOG_ERR "No SCMI shared memory for domain\n");
+        return false;
+    }
+
+    data = (struct scmi_shared_mem*)xzalloc_array(char, 256);
+    if ( !data )
+    {
+        printk(XENLOG_ERR "Could not allocate buffer for SCMI SHM\n");
+        return false;
+    }
+
+    flush_page_to_ram(page_to_mfn(current->domain->arch.scmi_base_pg), false);
+    res = access_guest_memory_by_ipa(current->domain,
+                                     current->domain->arch.scmi_base_ipa,
+                                     data, 256, false);
+    if ( res )
+    {
+        printk(XENLOG_ERR "Error reading guest memory %d\n", res);
+        goto err;
+    }
+
+    hdr = data->msg_header;
+
+    switch ( SCMI_HDR_PROTO_ID(data->msg_header) )
+    {
+    case SCMI_PROTOCOL_BASE:
+        handle_base_req(data);
+        break;
+    case SCMI_PROTOCOL_PERF:
+        handle_perf_req(data);
+        break;
+    default:
+        write_32(data->msg_payload, SCMI_ERR_SUPPORT);
+        data->length = sizeof(uint32_t) * 2;
+        break;
+    }
+
+    data->channel_status = SCMI_SHMEM_CHAN_STAT_CHANNEL_FREE;
+
+    res = access_guest_memory_by_ipa(current->domain,
+                                     current->domain->arch.scmi_base_ipa,
+                                     data, 256, true);
+
+    flush_page_to_ram(page_to_mfn(current->domain->arch.scmi_base_pg), false);
+    if ( res )
+    {
+        printk(XENLOG_ERR "Error writing guest memory %d\n", res);
+        goto err;
+    }
+
+    xfree(data);
```

Comment:
You free data whether error or not. Maybe something like:

err:
    xfree(data);
    return !res;

Reply:
Good idea, thanks